[x-plane] Update to 4.0.1, effectively new port

judemille opened the pull request:
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

judemille committed:
[xplane-sdk] New port

judemille commented:
Aware of build issues. This port is currently problematic on Mac too, from reports. Currently working to determine the best way to determine the target system type when building under vcpkg.

judemille commented:
Have noticed existing port `x-plane`. Unsure how to proceed, as that's not a particularly good name for the port. Regardless, this is a newer version, and with CMake targets. I'm currently working on dealing with the oddities of a prebuilt target.

judemille committed:
[xplane-sdk] This build should work.

judemille committed:
[xplane-sdk] updated versions. i forgor.

judemille committed:
[xplane-sdk] Build fixed for Mac.

Cheney-W commented:
@judemille 
I've checked the existing port `x-plane`, and it indeed shares the same upstream as the port you added. I believe we can proceed with the following steps now:
1. Rename the new port to `xplane-sdk-devel` to match the results on https://repology.org/project/xplane-sdk-devel/versions.
2. Set the old port `x-plane` to an empty port and make it dependent on the new `xplane-sdk-devel`.


@JavierMatosD What do you think?

judemille commented:
I find the `-devel` suffix unnecessary -- I'm not even sure why it's there on that AUR package, since it's not convention for Arch packages to have a `-devel` variant.

dg0yt commented:
AFAICS the upstream spelling is "X-Plane", so "x-plane" is a reasonable port name.
https://developer.x-plane.com/

judemille commented:
@dg0yt The main reason I picked `xplane-sdk` and not `x-plane` is since this port is the SDK for the flight simulator, and not the simulator itself. I could rename this to `x-plane-sdk`.

Cheney-W commented:
I think `x-plane-sdk` would be more accurate.

judemille committed:
[x-plane-sdk] xplane-sdk renamed to x-plane-sdk.

xplane-sdk has not yet been published, so this is okay.

judemille committed:
[x-plane-sdk] Renaming fixes.

judemille committed:
[x-plane-sdk] Clarify license.

The license seems to be a weird mix of the MIT, BSD 1-clause, and BSD
3-clause licenses. I can't work it out. It's permissive, but it's not
any one license.

judemille commented:
I don't like setting the license to `null`, but I legitimately can't work out what it is -- it most closely matches the BSD family, but it's not any BSD license known to me.

dg0yt commented:
> I don't like setting the license to `null`, but I legitimately can't work out what it is -- it most closely matches the BSD family, but it's not any BSD license known to me.
> 
> E: Maybe it's `BSD-Source-Code`?

You can try https://tools.spdx.org/app/check_license/, but double check the result.

If in doubt, `null` is fine, with the meaning it has in vcpkg.

judemille commented:
> You can try https://tools.spdx.org/app/check_license/, but double check the result.
> 
> If in doubt, `null` is fine, with the meaning it has in vcpkg.

That has no idea what I've given it, so I'll go with `null`.

Cheney-W commented on ports/x-plane-sdk/portfile.cmake:
```diff
@@ -0,0 +1,31 @@
+
+vcpkg_cmake_install()
+
+file(
+    INSTALL "${SOURCE_PATH}/license.txt"
+    DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}"
+    RENAME copyright
+)
```
```suggestion
vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/license.txt")
```

judemille replied:
Oh, neat. Thanks!

judemille committed:
[x-plane-sdk] Switch to vcpkg_install_copyright.

judemille commented:
I think this is ready for review now, unless anyone has any other concerns. As I understand it, the conversion of the `x-plane` port to an empty port depending on this one should be done in a separate PR. Is that correct?

Cheney-W commented on ports/x-plane-sdk/portfile.cmake:
```diff
@@ -0,0 +1,27 @@
+set(XPSDK_VERSION "401")
```
```suggestion
vcpkg_check_linkage(ONLY_STATIC_LIBRARY)

set(XPSDK_VERSION "401")
```

judemille replied:
The X-Plane SDK itself is actually only dynamically linked, though the libraries this port will download are only stubs (and in the case of Linux, there is no stub, so one must tolerate undefined symbols). I've configured the `xplm_cpp` target to be static, though I could set it to respect the vcpkg linking option.

Cheney-W replied:
I noticed that you configured the `xplm_cpp` target as static, so I was hoping you could add the above statement. This way, it can inform users during the installation of the port that it will be compiled as static manner.

judemille replied:
Makes sense.

Cheney-W commented on ports/x-plane-sdk/portfile.cmake:
```diff
@@ -0,0 +1,27 @@
+    SOURCE_PATH "${SOURCE_PATH}"
+)
+
+vcpkg_cmake_install()
```
```suggestion
vcpkg_cmake_install()
vcpkg_cmake_config_fixup(PACKAGE_NAME unofficial-x-plane-sdk)
```

judemille replied:
Does this actually do anything?

Cheney-W replied:
Before adding, the `share\unofficial-x-plane-sdk` file structure is as follows:
![image](https://github.com/microsoft/vcpkg/assets/38240633/f09ebfb7-16eb-400b-9a12-f5860d45755c)

After adding, the `share\unofficial-x-plane-sdk` file structure is as follows:
![image](https://github.com/microsoft/vcpkg/assets/38240633/2c9b4860-134f-45f0-9be5-75174afe60fb)

judemille replied:
I see. I'll commit that.

judemille committed:
[x-plane-sdk] Fixup config.

Cheney-W commented:
I think this PR should be fine, could you please click the ready for review button? Thank you!

judemille commented:
Already done lol

data-queue commented:
I'm not sure if we want to rename the port unless there is a strong reason to, as this would break users of the `x-plane` port.

judemille commented:
Given tools, and thus binaries are planned (or already? have not checked in a bit) to be put in vcpkg, I find the distinction between the actual simulator and its plugin library to be important.

judemille commented:
(pinging @data-queue in case no notification was sent for my response)

BillyONeal commented:
> Given tools, and thus binaries are planned (or already? have not checked in a bit) to be put in vcpkg, I find the distinction between the actual simulator and its plugin library to be important.

The issue is that we've already given the name `x-plane` to the SDK, and there are going to be a nonzero number of users who are broken with an attempted rename to something else.

Given that vcpkg is not an application deployment channel it seems unlikely that the actual simulator is going to be delivered this way, so I don't think we should be worrying about what happens with that.

That all said I'm kinda surprised they don't stomp on each other:

```console
PS D:\vcpkg> .\vcpkg.exe install x-plane x-plane-sdk --triplet x64-windows
Computing installation plan...
The following packages will be built and installed:
  * vcpkg-cmake:x64-windows -> 2023-05-04
  * vcpkg-cmake-config:x64-windows -> 2022-02-06#1
    x-plane:x64-windows -> 3.0.3#2
    x-plane-sdk:x64-windows -> 4.0.1
Additional packages (*) will be modified to complete this operation.
Detecting compiler hash for triplet x64-windows...
Restored 2 package(s) from C:\Users\bion\AppData\Local\vcpkg\archives in 49.5 ms. Use --debug to see more details.
Installing 1/4 vcpkg-cmake:x64-windows...
Elapsed time to handle vcpkg-cmake:x64-windows: 3.52 ms
Installing 2/4 vcpkg-cmake-config:x64-windows...
Elapsed time to handle vcpkg-cmake-config:x64-windows: 2.11 ms
Installing 3/4 x-plane:x64-windows...
Building x-plane:x64-windows...
-- Using cached XPSDK303.zip.
-- Extracting source D:/vcpkg-downloads/XPSDK303.zip
-- Using source at D:/vcpkg/buildtrees/x-plane/src/XPSDK303-d91b486681.clean
-- Installing: D:/vcpkg/packages/x-plane_x64-windows/share/x-plane/copyright
-- Performing post-build validation
Stored binaries in 1 destinations in 41.8 ms.
Elapsed time to handle x-plane:x64-windows: 279 ms
Installing 4/4 x-plane-sdk:x64-windows...
Building x-plane-sdk:x64-windows...
-- Note: x-plane-sdk only supports static library linkage. Building static library.
-- Downloading https://developer.x-plane.com/wp-content/plugins/code-sample-generation/sample_templates/XPSDK401.zip -> XPSDK401.zip...
-- Extracting source D:/vcpkg-downloads/XPSDK401.zip
-- Using source at D:/vcpkg/buildtrees/x-plane-sdk/src/XPSDK401-8960f386ff.clean
-- Configuring x64-windows
-- Building x64-windows-dbg
-- Building x64-windows-rel
-- Installing: D:/vcpkg/packages/x-plane-sdk_x64-windows/share/x-plane-sdk/copyright
-- Performing post-build validation
Stored binaries in 1 destinations in 137 ms.
Elapsed time to handle x-plane-sdk:x64-windows: 3.1 s
Total install time: 3.4 s
x-plane-sdk provides CMake targets:

    find_package(unofficial-x-plane-sdk CONFIG REQUIRED)
    target_link_libraries(main PRIVATE unofficial::x-plane-sdk::xplm unofficial::x-plane-sdk::xpwidgets unofficial::x-plane-sdk::xplm_cpp)
```

Which of these are callers expected to say in their .cpp files?

* `#include <x-plane-sdk/XPLM/XPLMDisplay.h>` (As proposed here)
* `#include <XPLM/XPLMDisplay.h>`
* `#include <XPLMDisplay.h>` (As in the existing `x-plane` port)

judemille commented:
With the current way CMake is set up, it should allow the latter:
`#include <XPLMDisplay.h>`

This is convention for X-Plane plugins.

BillyONeal commented:
> With the current way CMake is set up, it should allow the latter: `#include <XPLMDisplay.h>`
> 
> This is convention for X-Plane plugins.

Then I think the previous port is doing the correct thing to put them in all directly in `include`, since other build systems like msbuild are going to do/expect that rather than necessarily looking at CMake configs. But if this port were changed to do that, it doesn't coexist with the other one anymore. On the other hand, the status quo is that they don't actually coexist because given a source file that says `#include <XPLMDisplay.h>` you will get some mix of stuff from both ports, which seems undesirable.

I meant to post this screenshot before:

![diff screenshot of x-plane vs x-plane-sdk](https://github.com/microsoft/vcpkg/assets/1544943/ca617ce1-9e9e-4050-8b1b-48e16ea34422)

As a result I think the right thing is:

1. Change this to just update the existing `x-plane` port to the new version
2. Install the headers directly into `include`

judemille commented:
Fair enough. Given the whole thing of port compatibility I guess it makes sense to just rename this to `x-plane` and then move the headers.

judemille committed:
Rename x-plane-sdk to x-plane.

Includes also moved to include root.

Signed-off-by: Julia DeMille <me@jdemille.com>

Cheney-W commented on ports/x-plane/vcpkg.json:
```diff
@@ -1,8 +1,16 @@
   "name": "x-plane",
-  "version": "3.0.3",
-  "port-version": 2,
-  "description": "The X-Plane Plugin SDK",
-  "homepage": "https://developer.x-plane.com/sdk/",
```
Please do not delete the description and homepage?

judemille replied:
Oops. My mistake, since I just moved my manifest in.

Cheney-W commented:
About `ports/x-plane/portfile.cmake`, please do not modify the line breaks throughout the entire text, as this will cause the modified diff to increase.

Cheney-W commented:
Please use CRLF in the portfile.cmake instead of LF, because it originally used CRLF.

judemille committed:
Forgot to format and update versions.

Signed-off-by: Julia DeMille <me@jdemille.com>

judemille commented:
@Cheney-W Fixed.

Cheney-W commented:
Is this PR ready for review?
